src/main/scala/org/apache/spark/deploy/armada/submit/ArmadaClientApplication.scala

```diff
@@ -139,25 +143,135 @@ private[spark] class ArmadaClientApplication extends SparkApplication {
+      throw new IllegalArgumentException(
+        "Either node selectors or gang scheduling node uniformity label must be set."
+      )
+    }
+  }
+
+  private def submitArmadaJob(
+      armadaClient: ArmadaClient,
+      clientArguments: ClientArguments,
       conf: SparkConf
-  ): Seq[Container] = {
-    (0 until numExecutors).map {
-      getExecutorContainer(_, driverServiceName, conf)
+  ): (String, Seq[String]) = {
+    val queue               = conf.get(ARMADA_JOB_QUEUE)
+    val jobSetId            = conf.get(ARMADA_JOB_SET_ID)
+    val (driver, executors) = newSparkJobSubmitRequestItems(clientArguments, conf)
+
+    val driverResponse = armadaClient.submitJobs(queue, jobSetId, Seq(driver))
+    val driverJobId    = driverResponse.jobResponseItems.head.jobId
+    log(
+      s"Submitted driver job with ID: $driverJobId, Error: ${driverResponse.jobResponseItems.head.error}"
+    )
+
+    val executorsResponse = armadaClient.submitJobs(queue, jobSetId, executors)
+    val executorJobIds = executorsResponse.jobResponseItems.map(item => {
+      log(
+        s"Submitted executor job with ID: ${item.jobId}, Error: ${item.error}"
+      )
+      item.jobId
+    })
+
+    (driverJobId, executorJobIds)
+  }
+
+  private def newSparkJobSubmitRequestItems(
+      clientArguments: ClientArguments,
+      conf: SparkConf
+  ): (api.submit.JobSubmitRequestItem, Seq[api.submit.JobSubmitRequestItem]) = {
+
+    val namespace = conf.get(ARMADA_SPARK_JOB_NAMESPACE)
+    val priority  = conf.get(ARMADA_SPARK_JOB_PRIORITY)
+
+    val primaryResource = clientArguments.mainAppResource match {
+      case JavaMainAppResource(Some(resource)) => Seq(resource)
+      case PythonMainAppResource(resource)     => Seq(resource)
+      case RMainAppResource(resource)          => Seq(resource)
+      case _                                   => Seq()
     }
+
+    val master = conf.get("spark.master")
+    val internalMaster =
+      if (conf.get(ARMADA_SERVER_INTERNAL_URL).nonEmpty)
+        s"local://armada://${conf.get(ARMADA_SERVER_INTERNAL_URL)}"
+      else {
+        master
+      }
+
+    val image         = conf.get("spark.kubernetes.container.image")
+    val executorCount = SchedulerBackendUtils.getInitialTargetExecutorNumber(conf)
+
+    val configGenerator = new ConfigGenerator("armada-spark-config", conf)
+    val gangAnnotations = GangSchedulingAnnotations(
+      None,
+      1 + executorCount,
+      conf.get(ARMADA_JOB_GANG_SCHEDULING_NODE_UNIFORMITY)
+    )
+    val annotations  = configGenerator.getAnnotations ++ gangAnnotations
+    val globalLabels = commaSeparatedLabelsToMap(conf.get(ARMADA_SPARK_POD_LABELS))
+    val driverLabels =
+      globalLabels ++ commaSeparatedLabelsToMap(conf.get(ARMADA_SPARK_DRIVER_LABELS))
+    val nodeSelectors = commaSeparatedLabelsToMap(conf.get(ARMADA_JOB_NODE_SELECTORS))
+
+    val driverServiceName = conf.get(SPARK_DRIVER_SERVICE_NAME_PREFIX) + Utils.randAlphanum()
```

Comment:
Spark Kubernetes uses

    val randomServiceId = KubernetesUtils.uniqueID(clock = clock)
    s"spark-$randomServiceId-driver-svc"

Reply:
I'd prefer to mimic Spark Kubernetes behaviour / reuse existing Spark Kubernetes code as much as possible. But it's up to your and Armada team's judgement.

Reply:
We do not import the Spark k8s dependency.

You also mentioned it for another thing.

Should we import it so we can reuse this logic and the other one?